### PR TITLE
nrf_wifi: Add binary patch file

### PR DESCRIPTION
**Pull request by krish2718**
This remove C header based patch file and adds binary patch file.

Implements SHEL-2110.

### CODE REVIEW

**Review by udaynordic — APPROVED** [review]

**Review by rado17 — APPROVED** [review]

**Review by VivekUppunda — APPROVED** [review]

File: nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * @brief nRF70 patch version.
+ */
+#ifndef _PATCH_VERSION_H
+#define _PATCH_VERSION_H
+
+#include "pack_def.h"
+
+#ifndef BIT
+#define BIT(x) (1 << (x))
+#endif
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h:17]:
```suggestion
#endif /* BIT */
```

File: nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	NRF70_IMAGE_UMAC_PRI = 0,
+	NRF70_IMAGE_UMAC_SEC,
+	NRF70_IMAGE_LMAC_PRI,
+	NRF70_IMAGE_LMAC_SEC,
+};
+
+struct nrf70_fw_image {
+	unsigned int type;
+	unsigned int len;
+	/* Data follows */
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h:38]: Is it better to have this as a 0 byte array?

File: nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	unsigned int feature_flags;
+} __NRF_WIFI_PKD;
+
+
+#define RPU_VERSION         (1)
+#define RPU_MAJOR_VERSION   (2)
+#define RPU_MINOR_VERSION   (9)
+#define RPU_EXTRA_VERSION   (0)
+
+#endif /* _PATCH_VERSION_H */
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h:53]:
```suggestion
#endif /* _PATCH_INFO_H */
```

File: nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+} __NRF_WIFI_PKD;
+
+struct nrf70_fw_image_info {
+	unsigned int num_images;
+	unsigned int version;
+	unsigned int feature_flags;
+} __NRF_WIFI_PKD;
+
+
+#define RPU_VERSION         (1)
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h:48]:
```suggestion
#define RPU_FAMILY         (1)
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 5]
```diff
@@ -24,6 +24,177 @@
...
+
+	return 0;
+}
+
+enum nrf_wifi_status nrf_wifi_fmac_fw_parse(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+					   const void *fw_data,
+					   unsigned int fw_size,
+					   struct nrf_wifi_fmac_fw_info *fw_info)
+{
+	struct nrf70_fw_image_info  *info = (struct nrf70_fw_image_info *)fw_data;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:103]:
```suggestion
	struct nrf70_fw_image_info *info = (struct nrf70_fw_image_info *)fw_data;
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 6]
```diff
@@ -24,6 +24,177 @@
...
+	int offset = 0;
+	int image_id;
+
+	if (!fw_data || !fw_size || !fw_info) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Invalid parameters");
+		return -1;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:110]:
```suggestion
		return NRF_WIFI_STATUS_FAIL;
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 7]
```diff
@@ -24,6 +24,177 @@
...
+	}
+
+	if (fw_size < sizeof(struct nrf70_fw_image_info)) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Invalid fw_size: %d, minimum size: %d",
+			fw_size, sizeof(struct nrf70_fw_image_info));
+		return -1;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:117]:
```suggestion
		return NRF_WIFI_STATUS_FAIL;
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 8]
```diff
@@ -24,6 +24,177 @@
...
+	}
+
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"num_images: %d", info->num_images);
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"version: 0x%x", info->version);
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"feature_flags: %d", info->feature_flags);
+
+	if (info->num_images != NRF_WIFI_PATCH_NUM_IMAGES) {
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:127]: Is this check always true? As in can't we have only UMAC or LMAC patch?

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 9]
```diff
@@ -24,6 +24,177 @@
...
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Invalid number of images, expected %d, got %d",
+			NRF_WIFI_PATCH_NUM_IMAGES, info->num_images);
+		return -1;
+	}
+
+	if (nrf_wifi_patch_version_compat(fmac_dev_ctx, info->version) != 0) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible patch version");
+		return -1;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:137]:
```suggestion
		return NRF_WIFI_STATUS_FAIL;
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 10]
```diff
@@ -24,6 +24,177 @@
...
+	}
+
+	if (nrf_wifi_patch_feature_flags_compat(fmac_dev_ctx, info->feature_flags) != 0) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible feature flags");
+		return -1;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:143]:
```suggestion
		return NRF_WIFI_STATUS_FAIL;
```

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 11]
```diff
@@ -24,6 +24,177 @@
...
+	}
+
+	offset = sizeof(struct nrf70_fw_image_info);
+
+
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:148]: Extra newline

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 12]
```diff
@@ -24,6 +24,177 @@
...
+	return 0;
+}
+
+enum nrf_wifi_status nrf_wifi_fmac_fw_parse(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+					   const void *fw_data,
+					   unsigned int fw_size,
+					   struct nrf_wifi_fmac_fw_info *fw_info)
+{
+	struct nrf70_fw_image_info  *info = (struct nrf70_fw_image_info *)fw_data;
+	int offset = 0;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:104]: This initialization does not seem necessary.

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c [hunk 13]
```diff
@@ -24,6 +24,177 @@
...
+		default:
+			nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+				"Invalid image id: %d", image_id);
+			break;
+		}
+
+		offset += sizeof(struct nrf70_fw_image) + image->len;
+	}
+
+	return 0;
```
Comment (sachinthegreen) [inline nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:195]:
```suggestion
	return NRF_STATUS_SUCCESS;
```

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h (hunk 2 above)
Comment (krish2718) [reply nrf_wifi/fw_if/umac_if/inc/fw/patch_info.h:38]: That isn't allowed in Zephyr as its non-standard, it has to be flexible array, but as we don't use that, just removed.

File: nrf_wifi/fw_if/umac_if/src/fmac_api_common.c (hunk 8 above)
Comment (krish2718) [reply nrf_wifi/fw_if/umac_if/src/fmac_api_common.c:127]: The `NRF_WIFI_PATCH_NUM_IMAGES` is defined in the `patch_info.h` and is controlled by RPU. We just do a sanity check against the binary.

**Review by sachinthegreen — APPROVED** [review]

**Pull request merged**